### PR TITLE
Document testing in `pulumi/pulumi`

### PR DESCRIPTION
**Pull request by lunaris**
This commit begins the process of documenting how we test Pulumi, from unit tests to language conformance. It's nowhere near complete yet but the various sections have been stubbed out and anything that already exists has been ported over and linked to other new pages where possible. This commit is a good example of where we can start to take advantage of Sphinx's rich cross-references, placing e.g. `README.md` files close to the things they document (such as
`cmd/pulumi-test-language/README.md`) while folding them into a larger section on testing using Sphinx's tables of contents (TOCs).

### CODE REVIEW

**Comment by lunaris** [comment]
New pages hanging off https://pulumi-developer-docs--17161.org.readthedocs.build/17161/docs/architecture/testing/README.html

File: cmd/pulumi-test-language/README.md [hunk 1]
```diff
@@ -1,63 +1,134 @@
...
+
+For each test run and language, then:
+
+* If the test requires one or more providers, SDKs are generated from the
+  relevant test providers, exercising SDK generation for the given language.
+* The PCL program is converted into a program in the target language, exercising
+  program generation for the given language.
+* The generated program is executed using a compiled language host running as a
+  separate process and the assertions are checked, exercising program execution
+  for the given language.
```
Comment (Frassle) [inline cmd/pulumi-test-language/README.md:26]: Worth noting all the generated code is "snapshot" tested to ensure it doesn't change unexpectedly.

File: cmd/pulumi-test-language/README.md [hunk 2]
```diff
@@ -1,63 +1,134 @@
...
+conformance tests are defined in
+<gh-file:pulumi#cmd/pulumi-test-language/testdata>.
+<gh-file:pulumi#cmd/pulumi-test-language/tests.go> then references these
+programs and defines the assertions to be made about each. Tests are categorised
+as follows:
+
+* *L1 tests* are those which *do not* exercise provider code paths and use only
+  the most basic of features (e.g. stack outputs).
+* *L2 tests* are those which *do* exercise provider code paths and use things
+  such as custom resources, function invocations, and so on.
```
Comment (Frassle) [inline cmd/pulumi-test-language/README.md:41]: We are going to add L3 tests, it's probably worth documenting it now. They're the set of tests that need "apply" features that we don't expect to support in YAML.

File: cmd/pulumi-test-language/README.md [hunk 3]
```diff
@@ -1,63 +1,134 @@
...
+
+Each language defines a Go test function (the *language test host*) responsible
```
Comment (Frassle) [inline cmd/pulumi-test-language/README.md:43]: Strictly speaking doesn't have to be a Go test.

**Pull request merged**